docs/proposals/20221205-cpu-schedule-latency.md

```diff
@@ -0,0 +1,240 @@
+
+## Motivation
+
+After a linux task(thread) enters the *TASK_RUNNING* state from any state, (such as the I/O operation is completed), it is first placed in the *Runnable* stage, waiting for the allocation of CPU time slices in the runqueue, at this time all the conditions for the task to run have been met. Scheduling latency refers to the period from when a task is ready to run (enters the CPU's runqueue) to when it is actually executed (obtains the execution right of the CPU). This time span depends on whether other tasks are running on the current CPU and the number of tasks queued to run. It is generally believed that the more tasks queued, the longer the scheduling latency may be. So this is also a metric to indirectly measure the CPU load, which is obtained by calculating the number of tasks on the runqueue at each time. In a colocation scenario, other tasks or run queue squeeze will cause the scheduling latency of the expected application process (such as high-priority online tasks) to increase. It can be considered that CPU scheduling latency should be a powerful tool for Interference Detection. 
+
+*eBPF* is a revolutionary technology with origins in the Linux kernel that can run sandboxed programs in an operating system kernel. Today, eBPF is used extensively to drive a wide variety of use cases including extracting fine-grained security observability data at low overhead in cloud native environments. By tracing schedule events using eBPF, we can compute CPU schedule latency easily. 
+
+*cilium/ebpf* is a pure Go library that provides utilities for loading, compiling, and debugging eBPF programs. It has minimal external dependencies and is intended to be used in long running processes.
+
+### Goals
+
+- Define structures, BPF Maps, and working mechanism for eBPF CPU schedule latency collector. 
+- Explain workflow of Koordinator and CPU schedule latency collector. 
+
+### Non-Goals/Future Work
+
+- Design external metrics collector framework to run metric collection without Koordlet invasion.
+- Define CRD or API for the collector to upload CPU schedule latency. 
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a cluster administrator, I want to make use of Koordinator's Interference Detection feature to gain better cluster performance. 
+
+### Design
+
+#### Computation
+
+The CPU Schedule Latency is defined as below. Tracepoint `tp/sched/sched_wakeup` and `tp/sched/sched_wakeup_new` indicate one task is enqueued into `runqueue`.  And tracepoint `tp/sched/sched_switch` means one task finishes context switch. 
+
+![cpu-schedule-latency](../images/cpu-schedule-latency.jpg)
+
+Notice that there might be many tasks and multiple times of context switch per task in one cgroup. So to compute CPU Schedule Latency on cgroup dimension, we sum all tasks' latency and record the number of latency reported by collector. The final CPU Schedule Latency is defined as `sum(latency)/num(latency)`. 
```

Comment:
not only average, but also percentile values.

Reply:
the `sum` or `count` value is also important when we want to check the accumulated latency of a cgroup

Reply:
So we can use histogram to handle such requirements. ✍️